openadapt/evaluators/evaluator.py

```diff
@@ -0,0 +1,183 @@
+from typing import Tuple
+
+from loguru import logger
+
+from openadapt import utils
+from openadapt.evaluators import fixtures
+from openadapt.models import KeyAction, MouseAction, Window
+
+LOG_LEVEL = "DEBUG"
+DEFAULT_MAX_SCREEN_SIZE = (1920, 1080)
+DEFAULT_MAX_INPUT_SIZE = 1024
+DEFAULT_MAX_TOKENS = 1024
+
+utils.configure_logging(logger, LOG_LEVEL)
+
+
+class BaseEvaluator:
+    """Base class for all evaluations"""
+
+    def __init__(
+        self,
+        model_name: str = "gpt2",
+        max_input_size: int = DEFAULT_MAX_INPUT_SIZE,
+        max_tokens: int = DEFAULT_MAX_TOKENS,
+        max_screen_size: Tuple[int, int] = DEFAULT_MAX_SCREEN_SIZE,
+    ):
+        self.model_name = model_name
+        self.max_input_size = max_input_size
+        self.max_tokens = max_tokens
+        self.max_screen_size = max_screen_size
+
+    def init_model(self):
+        """Initialize the model. This must be adapted for each model, given attribute model name"""
+        raise NotImplementedError
+
+    def get_completion(
+        self,
+        prompt: str,
+    ) -> str:
+        """Get the completion for a given prompt"""
+        raise NotImplementedError
+
+    def build_prompt(
+        self, ref_window: Window, action: KeyAction | MouseAction, active_window: Window
+    ):
+        """Build the prompt for the model. This must be adapted for each model.
+
+        Args:
+            ref_window (Window): reference window
+            action (KeyAction | MouseAction): action to be performed
+            active_window (Window): active window
+
+        Returns:
+            str: prompt for the model
+        """
+        prompt = (
+            f"{ref_window.dict()=} {action.dict()=} {active_window=} # Provide valid"
+            " Python3 code containing the action dicts by completing the following,"
+            " and nothing else: active_action_dict="
+        )
+        return prompt
+
+    def evaluate(self) -> float:
+        """Evaluate the model completion on a set of fixtures
+
+        Returns:
+            float: average score
+        """
+        self.init_model()
+        # TODO: run evaluation on different fixtures
+        ref_window, action, active_window = fixtures.generate_single_mouse()
+        prompt = self.build_prompt(ref_window, action, active_window)
+        completion = self.get_completion(prompt)
+        score = self._score_completion(completion, active_window, action)
+        logger.info(f"Average score is {score=}")
+        return score
+
+    def _is_valid_position(
+        self, position: tuple[int, int], active_window: Window
+    ) -> bool:
+        if (
+            position[0] < active_window.left
+            or position[0] > active_window.left + active_window.width
+        ):
+            return False
+        if (
+            position[1] < active_window.top
+            or position[1] > active_window.top + active_window.height
+        ):
+            return False
+        return True
```

Comment:
Now the evaluation criteria is simplified:

- Any action which is of valid type
- If the action is of type key press, just check the correct key
- If the action is of type mouse movement/click, verify if the position is withint the boundary of the active window.
- Refer to https://mldsai.slack.com/archives/C05JCPC5HAS/p1691753011770869, it is as of now really hard or impossible to identify which element of the reference window is clicked. Even the `state` does not contain much data and such data has to be tremendously augmented to achieve the deteciton of "focus" area. We can discuss this point further separately.

Reply:
cc: @abrichr

Reply:
Sum up: 
- Test again with other type of application other than Web Browser and Citrix. It is may be still possible to identify the active elements on the active windows if the windows is not of type Web Browser or Citrix.

To be updated soon.